crates/bevy_ecs/src/world/mod.rs

```diff
@@ -0,0 +1,819 @@
+    /// ```
+    pub fn resource_scope<T: Component, U>(
+        &mut self,
+        f: impl FnOnce(&mut T, &mut World) -> U,
+    ) -> U {
+        let mut resource = self
+            .remove_resource::<T>()
+            .expect("resource does not exist");
+        let result = f(&mut resource, self);
+        self.insert_resource(resource);
+        result
+    }
+
+    /// # Safety
+    /// `component_id` must be assigned to a component of type T
+    #[inline]
+    pub(crate) unsafe fn get_resource_with_id<T: 'static>(
+        &self,
+        component_id: ComponentId,
+    ) -> Option<&T> {
+        let column = self.get_populated_resource_column(component_id)?;
+        Some(&*column.get_ptr().as_ptr().cast::<T>())
+    }
+
+    /// # Safety
+    /// `component_id` must be assigned to a component of type T.
+    /// Caller must ensure this doesn't violate Rust mutability rules for the given resource.
+    #[inline]
+    pub(crate) unsafe fn get_resource_unchecked_mut_with_id<T>(
+        &self,
+        component_id: ComponentId,
+    ) -> Option<Mut<'_, T>> {
+        let column = self.get_populated_resource_column(component_id)?;
+        Some(Mut {
+            value: &mut *column.get_ptr().as_ptr().cast::<T>(),
+            flags: &mut *column.get_flags_mut_ptr(),
+        })
+    }
+
+    /// # Safety
+    /// `component_id` must be assigned to a component of type T
+    #[inline]
+    pub(crate) unsafe fn get_non_send_with_id<T: 'static>(
+        &self,
+        component_id: ComponentId,
+    ) -> Option<&T> {
+        self.validate_non_send_access::<T>();
+        self.get_resource_with_id(component_id)
+    }
+
+    /// # Safety
+    /// `component_id` must be assigned to a component of type T.
+    /// Caller must ensure this doesn't violate Rust mutability rules for the given resource.
+    #[inline]
+    pub(crate) unsafe fn get_non_send_unchecked_mut_with_id<T: 'static>(
+        &self,
+        component_id: ComponentId,
+    ) -> Option<Mut<'_, T>> {
+        self.validate_non_send_access::<T>();
+        self.get_resource_unchecked_mut_with_id(component_id)
+    }
+
+    /// # Safety
+    /// `component_id` must be valid and correspond to a resource component of type T
+    #[inline]
+    unsafe fn insert_resource_with_id<T>(&mut self, component_id: ComponentId, mut value: T) {
+        let column = self.initialize_resource_internal(component_id);
+        if column.is_empty() {
+            // SAFE: column is of type T and has been allocated above
+            let data = (&mut value as *mut T).cast::<u8>();
+            // SAFE: new location is immediately written to below
+            let row = column.push_uninit();
+            // SAFE: index was just allocated above
+            column.set_unchecked(row, data);
+            column
+                .get_flags_unchecked_mut(row)
+                .set(ComponentFlags::ADDED, true);
+            std::mem::forget(value);
```

Comment:
```suggestion
            std::mem::forget(value);
            column
                .get_flags_unchecked_mut(row)
                .set(ComponentFlags::ADDED, true);
```
Just in case `get_flags_unchecked_mut` or `set` panics for whatever reason.